core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -1259,15 +1255,33 @@ public RelBuilder projectPlus(RexNode... nodes) {
     return projectPlus(ImmutableList.copyOf(nodes));
   }
 
-  /** Creates a {@link Project} of all original fields, plus the given list of
+  /** Creates a {@link Project} of all original fields, plus the given
    * expressions. */
   public RelBuilder projectPlus(Iterable<RexNode> nodes) {
     final ImmutableList.Builder<RexNode> builder = ImmutableList.builder();
     return project(builder.addAll(fields()).addAll(nodes).build());
   }
 
-  /** Creates a {@link Project} of the given list
-   * of expressions, using the given names.
+  /** Creates a {@link Project} of all original fields, except the given
+   * expressions. */
+  public RelBuilder projectExcept(RexNode... expressions) {
+    return projectExcept(ImmutableList.copyOf(expressions));
+  }
+
+  /** Creates a {@link Project} of all original fields, except the given
+   * expressions. */
+  public RelBuilder projectExcept(Iterable<RexNode> expressions) {
+    List<RexNode> allExpressions = new ArrayList<>(fields());
+    for (RexNode excludeExp : expressions) {
+      if (!allExpressions.remove(excludeExp)) {
+        throw RESOURCE.expressionNotFound(excludeExp.toString()).ex();
```

Comment:
I think we should still check the redundant same digest inputs, for example, if we `projectExcept(['a', 'a'])` from `a, b, c`, it should not throws `expression not found` but `Redundant expression 'a' to remove` (or something similar) which is more readable.

Reply:
We can avoid the `redundant expression` exception by forcing the caller to pass a `Set` instead of an `Iterable`. If we are so concerned about duplicates then maybe this is a good alternative. Otherwise I can make the check and throw a new exception I have no strong preference over the two options so let me know what you prefer.

Reply:
+1 to pass a `Set`

Reply:
-1 to pass a `Set`, this would just make the interface complex, use can pass in any expression collections they like, what we need to do is to give the user-friendly validation.

Reply:
The thing is that it makes no sense that `expressions` contains duplicates. It's better to have a way to let users know this.

Reply:
My vote would be to throw IllegalArgumentException, not throw RESOURCE.something. This is an API error, not a user error.

Don't make too much effort to accommodate people who are using the API incorrectly. Make the specification as simple as possible, and if people use it wrong, throw. 

The simplest thing is probably `Collections.removeAll(Collection)`. Keep the argument as type `Iterable`, but cast or convert it to a collection.

Reply:
The javadoc should say what happens if the expressions do not exist. It's part of the contract.